ribbon/src/test/java/feign/ribbon/RibbonClientCheckForIOExceptionTest.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2013 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feign.ribbon;
+
+import java.io.IOException;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+public class RibbonClientCheckForIOExceptionTest {
+
+    private Throwable throwable;
+
+    @Before
+    public void setUp() {
+
+        throwable = null;
+    }
+
+    @Test(expected = IOException.class)
+    public void assureThatIOExceptionIsFound() throws IOException {
+
+        givenAThrowableThatIsAnIOException();
+        whenCheckForIOExceptionIsCalled();
+    }
+
+    @Test(expected = IOException.class)
+    public void assureThatNestedIOExceptionIsFound() throws IOException {
+
+        givenAThrowableThatContainsAnIOException();
+        whenCheckForIOExceptionIsCalled();
+    }
+
+    @Test
+    public void assureNothingHappensIfThereIsNoIOException() throws IOException {
+
+        givenAThrowableWithoutIOException();
+        whenCheckForIOExceptionIsCalled();
+    }
+
+    @Test
+    public void assureNothingHappensIfThrowableIsNull() throws IOException {
+
+        givenThrowableIsNull();
+        whenCheckForIOExceptionIsCalled();
+    }
+
+    @Test
+    public void assureCyclesWillNotEndInAnEndlessLoop() throws IOException {
+
+        givenAThrowableWithCycles();
+        whenCheckForIOExceptionIsCalled();
+    }
+
+    private void givenAThrowableWithCycles() {
```

Comment:
since this is only about ribbon, make the exceptions concrete. ex always start with a clientException (unless instantiating that is somehow difficult)